### ISSUE TITLE
Problem with certain middlewares

### ISSUE THREAD

**Issue by patrickdappollonio** [pressly/cji#2]
Hello,

I was trying to use an authentication middleware in my project together with Cji (which I have to say, make everything less painful regarding chaining middlewares on Goji projects) but I cannot seem to get it working...

The middleware I was trying to use was [httpauth](https://github.com/goji/httpauth) from Goji itself. From what I can see, the problem is at this line `fn = mw(http.HandlerFunc(newFn)).(http.HandlerFunc)` where Cji assumes that the function could be asserted as an `http.HandlerFunc`.

Here's how I declared the route:

```go
goji.Get("/panel/categories", cji.Use(boltHandler, httpauth.SimpleBasicAuth("test", "pass")).On(handler))
```

Any ideas?

```
2015/03/10 18:59:35.862987 [COMPUTER/47zK9KnkTB-000012] Returning 304 in 1ms
2015/03/10 18:59:42.655350 [COMPUTER/47zK9KnkTB-000013] Started GET "/panel/categories" from 127.0.0.1:54665
2015/03/10 18:59:42.689350 [COMPUTER/47zK9KnkTB-000013] panic: interface conversion: http.Handler is httpauth.basicAuth, not http.HandlerFunc
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/middleware/recoverer.go:24 (0x5075b3)
        func.005: debug.PrintStack()
c:/go/src/runtime/asm_amd64.s:401 (0x42702c)
        call16: CALLFN(·call16, 16)
c:/go/src/runtime/panic.go:387 (0x4138cf)
        gopanic: reflectcall(unsafe.Pointer(d.fn), deferArgs(d), uint32(d.siz),uint32(d.siz))
c:/go/src/runtime/iface.go:181 (0x40bf32)
        assertI2T: panic(&TypeAssertionError{*tab.inter.typ._string, *tab._type._string, *t._string, ""})
C:/Users/TestUser/Go/src/github.com/pressly/cji.go:80 (0x4487f9)
        func.003: fn = mw(http.HandlerFunc(newFn)).(http.HandlerFunc)
C:/Users/TestUser/Go/src/github.com/pressly/cji.go:72 (0x44861a)
        func.002: hf(c, ww, rr)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/test/project/middleware/boltdb/boltdb.go:15 (0x448ee7)
        func.001: h.ServeHTTP(w, r)
C:/Users/TestUser/Go/src/github.com/pressly/cji.go:84 (0x44882c)
        func.003: fn(w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/web.go:113 (0x48acbc)
        HandlerFunc.ServeHTTPC: h(c, w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/router.go:114 (0x488eeb)
        (*router).route: route.handler.ServeHTTPC(*c, w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/middleware.go:88 (0x48ada7)
        func.002: router.route(&cs.C, w, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/test/project/middleware/renderer/renderer.go:16 (0x449630)
        func.001: h.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/test/project/middleware/static/static.go:33 (0x449ac8)
        func.001: h.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/test/project/middleware/promote/promote.go:15 (0x4492a3)
        func.001: h.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/middleware/options.go:70 (0x5074c5)
        func.003: h.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/middleware/recoverer.go:29 (0x507725)
        func.006: h.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/middleware/logger.go:35 (0x507269)
        func.001: h.ServeHTTP(lw, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/middleware/request_id.go:68 (0x507a4c)
        func.007: h.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1265 (0x462098)
        HandlerFunc.ServeHTTP: f(w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/middleware.go:46 (0x4857a9)
        (*cStack).ServeHTTP: s.m.ServeHTTP(w, r)
C:/Users/TestUser/Go/src/github.com/zenazn/goji/web/mux.go:65 (0x486c57)
        (*Mux).ServeHTTP: stack.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1541 (0x463834)
        (*ServeMux).ServeHTTP: h.ServeHTTP(w, r)
c:/go/src/net/http/server.go:1703 (0x4641c1)
        serverHandler.ServeHTTP: handler.ServeHTTP(rw, req)
c:/go/src/net/http/server.go:1204 (0x461bce)
        (*conn).serve: serverHandler{c.server}.ServeHTTP(w, w.req)
c:/go/src/runtime/asm_amd64.s:2232 (0x4291a1)
        goexit:
2015/03/10 18:59:42.722351 [COMPUTER/47zK9KnkTB-000013] Returning 500 in 67.0
013ms
```

**Comment by elithrar** [comment]
Note that `httpauth.SimpleBasicAuth` returns a  `func(http.Handler) http.Handler` which *should* be handled by the type switch here in cji: https://github.com/pressly/cji/blob/master/cji.go#L24

Do you have a minimal, compilable test I can try - i.e. if you comment out `httpauth.SimpleBasicAuth` in the `cji.Use()` chain, does it work?

**Comment by patrickdappollonio** [comment]
@elithrar I thought the same thing: that will be covered with the statement in the switch case. It does, but once it does it says that the type assertion is wrong, because `http.Handler` is `httpauth.basicAuth`, but not `http.HandlerFunc`.

If I comment the function called, the system keeps working as usual. That's why I can't figure it out. I've asked @pkieltyka over Slack and he told me that, probably, the solution was to wrap it in an `http.HandlerFunc` but given the fact that it supposed to be already a handler, there's no need for that. 

I'll try to create a short piece of code to show the problem.

**Comment by elithrar** [comment]
Ping @zenazn - who may understand the root cause better than I.

**Comment by zenazn** [comment]
The problem is https://github.com/pressly/cji/blob/master/cji.go#L74 — it should read:
```
fn = mw(http.HandlerFunc(newFn)).ServeHTTP
```
(Although I definitely think `wrap` should be refactored to deal in `http.Handler`'s instead of `http.HandlerFunc`'s, but that's a little out of scope for this issue in particular)

**Comment by patrickdappollonio** [comment]
I confirm the solution works. I've created my own copy of cji and modified the offending line. In the meantime, this should keep us going. Thank you, @zenazn !

**Comment by pkieltyka** [comment]
fixed in 51b0fbb7a97f15d996e20f6e7a44c51abf5f4194 - thanks all!

**Comment by pkieltyka** [comment]
@zenazn I also agree, wrap should be refactored to dealing with http.Handler's
